oeapkman: Package Manager and Toolbox for Enclaves on Linux (#4157)
openenclave/openenclave: samples/apkman/README.md

# oeapkman Sample

This sample demonstrates installing and using [sqlite](https://sqlite.org/index.html)

> database static library via `oeapkman`. `sqlite` is used to create, populate

Reviewer: ```suggestion
database static library via `oeapkman`. `sqlite` is used to create, populate,
```

and query an in-memory database within an enclave.

# oeapkman Overview

`oeapkman` is a tool for installing and using [Alpine Linux](https://www.alpinelinux.org/)
static libraries in enclaves. Currently, `oeapkman` is Linux only.

## Why Alpine Linux?

OE SDK uses [MUSL](https://musl.libc.org/) as its enclave C library.
Most Linux distributions use GLIBC as the native C library.

> MUSL is currently not 100% binarycompatible with GLIBC as explained here:

Reviewer 2: ```suggestion
MUSL is currently not 100% binary compatible with GLIBC as explained here:
```

- https://www.musl-libc.org/faq.html
- https://wiki.musl-libc.org/functional-differences-from-glibc.html

> Libraries that successfully compile with GLIBC may not always

Reviewer: ```suggestion
Libraries that are successfully compiled with GLIBC may not always be
```

Author: Retaining the original tense; seems cleaner.

> compile with MUSL. There are also runtime behavior differences.

Reviewer: ```suggestion
compiled with MUSL. There are also runtime behavior differences.
```

Author: Ditto.

> Unlike other Linux distributions, Alpine Linux uses MUSL as it  s native C library.

Reviewer 2: ```suggestion
Unlike other Linux distributions, Alpine Linux uses MUSL as its native C library.
```

> Unlike other Linux distributions, Alpine Linux uses MUSL as it  s native C library.

Reviewer: ```suggestion
Unlike other Linux distributions, Alpine Linux uses MUSL as its native C library.
```

It's command syntax closely mirrors that of `apk`.

Commands have the following syntax: `oeapkman command parameters`

The command `oeapkman help` prints a message describing all the commands and their usage.

#### Installing packages

The command `oeapkman add package` installs the given package.

For example, the `sqlite` static library can be installed thus:

```bash
$ oeapkman add sqlite-static
(1/1) Installing sqlite-static (3.35.5-r0)
OK: 414 MiB in 47 packages
```

Multiple packages can be specified to install them all at once.
```bash
$ oeapkman add sqlite-static zlib-static
(1/2) Installing sqlite-static (3.35.5-r0)
(2/2) Installing zlib-static (1.2.11-r3)
OK: 414 MiB in 48 packages
```

To develop enclaves using a static library, both the headers files for the library also need
to be installed. The header files for static libraries are often found in the corresponding
`-dev` package. The header files needed for using `sqlite` are available in the `sqlite-dev`
package.

```bash
$ oeapkman add sqlite-dev
(1/2) Installing sqlite-libs (3.35.5-r0)
(2/2) Installing sqlite-dev (3.35.5-r0)
OK: 414 MiB in 48 packages
```

### Searching for packages

> If you are unsure of the exact package name, but know what it is (e.g sqlite),

Reviewer 2: ```suggestion
If you are unsure of the exact package name, but know what it is (e.g., sqlite),
```

you can find the exact package name at the
[Search Packages By Name](https://pkgs.alpinelinux.org/packages) site.

For example, by specifying "sqlite*" in the name field, and choosing branch v3.14 and
arch x86_64, the list of sqlite packages is found thus:
https://pkgs.alpinelinux.org/packages?name=sqlite*&branch=v3.14&arch=x86_64

Typically packages names `-static` contain static libraries and therefore `sqlite-static`
is the package that must be installed.

The license for each package is also listed in the search results.

On the other hand, if you know the contents of a package, but don't know its name, you
can find the exact package name by searching by contents at the
[Search Packages By Contents](https://pkgs.alpinelinux.org/contents) site.

For example, searching for the header file `sqlite3.h` via
https://pkgs.alpinelinux.org/contents?file=sqlite3.h&path=&name=&branch=edge&arch=x86_64
reveals that `sqlite-dev` is the package to install for obtaining `sqlite` headers.

The relative path for the file is also listed. This is helpful for specifying the path

> to the header file to the compiler. See section below about `apkman root`.

Reviewer 2: ```suggestion
to the header file to the compiler. See the [apkman root](#oeapkman-root] section below.
```

Similarly, searching for `libsqlite*.a` via
https://pkgs.alpinelinux.org/contents?file=libsqlite*.a&path=&name=&branch=edge&arch=x86_64
reveals that the package of interest is `sqlite-static` and that the library `libsqlite3.a`
exists in `/usr/lib`. This relative path is useful for helping the linker locate the library
as described in the next section.

### oeapkman root

Under the hood, `oeapkman` maintains an Alpine Linux distribution whose root filesystem resides
at the location given by the command `oeapkman root`:

```bash
$ oeapkman root
/home/username/.oeapkman/alpine-fs
```

The static libraries and header files installed via `oeapkman` are installed at their usual

> locations within the Alpine Linux root filesystem. Headers files are typically found in

Reviewer 2: ```suggestion
locations within the Alpine Linux root filesystem. Header files are typically found in
```

    SQL_TEST(sqlite3_finalize(stmt));
    SQL_TEST(sqlite3_close(db));
```

As described earlier, `sqlite` library is installed using the `oeapkman` command
and the header files and static library paths are specified using the `ROOT` property
of `oeapkman` target.

[enclave/CMakeLists.txt](enclave/CMakeLists.txt)
```cmake
get_target_property(APKMAN_ROOT openenclave::oeapkman ROOT)

# For sqlite headers.
target_include_directories(
  enclave
  PRIVATE ${CMAKE_CURRENT_BINARY_DIR}
          # sqlite include path.
          ${APKMAN_ROOT}/usr/include)

target_link_libraries(
  enclave openenclave::oeenclave "-L  ${APKMAN_ROOT}/usr/lib" libsqlite3.a
  openenclave::oecrypto${OE_CRYPTO_LIB} openenclave::oelibc)
```

Similarly, in the Makefile based workflow, `oeapkman add sqlite-static sqlite-dev` and
`oeapkman root` commands are used to enable use of `sqlite` static library.

See [enclave/Makefile](enclave/Makefile)

## stubs

`sqlite` uses some functions that are not currently implemented by OE SDK.
- dlopen, dlclose, dlsym, dlerror
- fchmod, fchown
- mremap
- posix_fallocate
- readlink
- utimes

Since in-memory databases work without implementing these functions, these

> functions have been provided empty implementations (stubs).

Author: Good point.

# oeapkman Sample

This sample demonstrates installing and using the [sqlite](https://sqlite.org/index.html)
database static library via `oeapkman`. `sqlite` is used to create, populate,
and query an in-memory database within an enclave.

# oeapkman Overview

`oeapkman` is a tool for installing and using [Alpine Linux](https://www.alpinelinux.org/)

> static libraries in enclaves. Currently, `oeapkman` is Linux only.

Reviewer 2: and SGX only?
